data/xfce4-screensaver.1
.\" xfce4-screensaver manual page
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH XFCE-SCREENSAVER 1 "14 October 2018" "Xfce Desktop Environment"
.\" Please adjust this date whenever revising the manpage.
.\"
.SH "NAME"
xfce4-screensaver \- The Xfce Desktop Screensaver and Locker
.SH "SYNOPSIS"
.B xfce4-screensaver [OPTIONS]
.SH "DESCRIPTION"
Xfce Screensaver is the default screen saver and locker for the Xfce Desktop. It is designed with simplicity and security in mind. \fBxfce4-screensaver\fP integrates well with the Xfce desktop, and it supports User Switching and the ability to lock down configuration settings.
.SH "OPTIONS"
.TP
\fB\-\-display=DISPLAY\fR
X display to use.
.TP
.TP
\fB\-\-debug\fR
Enable debugging messages for \fBxfce4-screensaver\fP
.TP
\fB\-\-version\fR
Output version information and exit.
.TP
\fB\-h, \-\-help\fR
Print standard command line options.
.TP
\fB\-\-help\-all\fR
Print all command line options.
.P
This program also accepts the standard GTK options.
.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at: 
https://bugzilla.xfce.org
.SH "AUTHORS"
.SS xfce4-screensaver is maintained by:
.nf
Sean Davis <bluesabre@xfce.org>
Simon Steinbeiss <ochosi@xfce.org>
.fi
.SS This Man Page has been updated for the Xfce Desktop Environment by:
Sean Davis <bluesabre@xfce.org> (2018)
.SH "HISTORY"
.SS gnome-screensaver was written by William Jon McCann <mccann@jhu.edu>
.SS This manual page was originally written by Sven Arvidsson <sa@whiz.se> (2007)
.SH "SEE ALSO"
.SS Further information may also be available at: https://docs.xfce.org/
.P
.BR xfce4-screensaver-command (1),
.BR xfce4-screensaver-preferences (1),
.BR gtk-options (7)


NEWS
0.1.10 (beta)
=============
- Send DBus lock responses for xfce4-session support (bug #16522)
- Translation Update: Galician

0.1.9 (beta)
============
- Reimplement g_date_time_format_iso8601 to support GLib 2.50+
- Replace deprecated GTimeVal usage
- Rebuild windows on monitor reconfiguration
- Draw overlays during window reconstruction to protect screen
- Do not activate DPMS when screensaver is inactive (bug #16327)
- Better handling of multi-monitor and lid-close events (bug #16102)
- Update LINGUAS (bug #15949)
- Fix decimal properties when running through atof
- Return 1 on lock command failure (bug #15945)
- Rename 'Pictures folder' to 'Slideshow' (bug #15589)
- Raise NameError and TypeError (bug #15830)
- Fix float parsing error (bug #16295)
- Fix inhibitor proxying (bug #16356)
- Fix inhibitor listing in xfce4-screensaver-command (bug #16355)
- Add systemd sleep inhibitor (bug #15929)
- Fix dbus inhibition (bug #16365)
- dbus: Prevent overzealous activation (bug #16365)
- Translation Updates:
  Albanian, Belarusian, Czech, Finnish, Galician, Hebrew, Italian,
  Japanese, Korean, Malay, Norwegian Bokmål, Polish, Portuguese
  (Brazil), Serbian, Spanish, Swedish, Turkish

0.1.8 (beta)
============
- Features:
  - Optional screensaver inhibition for fullscreen apps
  - Blank screensaver now supports DPMS (Xfce #15216)
- Build Changes:
  - Add dependency on libwnck >= 3.20
- Translation Updates:
  Basque, Bulgarian, Catalan, Chinese (China), Chinese (Taiwan),
  Danish, Dutch, Finnish, French, Galician, German, Hungarian,
  Italian, Lithuanian, Polish, Portuguese, Portuguese (Brazil),
  Russian, Spanish

0.1.7 (beta)
============
- Features:
  - Added configuration tool for individual screensavers
  - Added support for on-screen keyboards
- Code Cleanup:
  - Drop libXxf86 dependency, no longer implemented or available
    (see https://github.com/mate-desktop/mate-screensaver/issues/199)
  - Removed unneeded login window construction code
  - Removed window shaking on failed login
  - Removed unused session-idle signal
  - Removed unused code from gs-manager/gs-window-x11
  - Simplified lock screen code
  - Simplified screensaver activation and locking code
- General:
  - Improved wallpaper drawing at 2x scale
  - Migrated xfce4-screensaver-command to GDBus
  - Fixed confusion between saver and lock timeouts
- Translation Updates:
  Armenian (Armenia), Belarusian, Bulgarian, Catalan, Chinese (China),
  Chinese (Taiwan), Czech, Danish, Dutch, Finnish, French, Galician,
  German, Hungarian, Italian, Lithuanian, Norwegian Bokmål, Polish,
  Portuguese, Portuguese (Brazil), Russian, Spanish, Turkish

0.1.6 (beta)
============
- Bug Fixes:
  - Fixed flickering in the password dialog
  - Fixed password dialog appearance, visual artifacts with various
    themes are resolved
  - Remove reference to pkg-config file (.pc) (Xfce #15597)
- Code Cleanup:
  - Cleaned up kdb-indicator logic
  - Consolidated debug function calls
  - Moved job theme processing out of gs-manager
  - Removed drag-and-drop theme installation
  - Removed placeholder widgets from the glade template
  - Removed unused callbacks
  - Simplified handling of user preferences
- Translation Updates:
  Armenian (Armenia), Hebrew, Hungarian, Malay, Portuguese,
  Portuguese (Brazil), Russian

0.1.5 (beta)
============
- Features:
  - Added support for elogind (Xfce #15243)
  - Added support for profile images from AccountsService (Xfce #15235)
- General:
  - Removed a lot of unused code (2,700+ lines)
  - Simplified monitor-handling logic
  - More consistent spacing in makefiles
- Bug Fixes:
  - Added categories to display in the Settings Manager (Xfce #15507)
  - Added a fallback idle timer for when X11 screensaver doesn't notify
  - Fixed warning during configure stage (Xfce #15552)
  - Fixed Inhibitor lock (Xfce #15551)
  - Hid screensaver themes from 'Open with' options (Xfce #15531)
  - Made numerous cppcheck, cpplint, and Coverity Scan fixes
    - Constant variables guarding dead code
    - Dereferencing before/after null
    - Unchecked return values
    - Unused variables
    - Memory leaks
- Translation Updates:
  Albanian, Basque, Belarusian, Bulgarian, Catalan, Chinese (China),
  Chinese (Taiwan), Czech, Danish, Dutch, English (Australia),
  French, German, Hebrew, Hungarian, Icelandic, Italian, Korean,
  Lithuanian, Malay, Polish, Portuguese (Brazil), Serbian, Slovak,
  Spanish, Swedish, Turkish

0.1.4 (beta)
============
- Features:
  - Added support for locking on suspend (Xfce #15001)
- Bug Fixes:
  - Fix password entry after stealing focus (Xfce #15042)
- General:
  - Faster screen locking when activated
  - Faster renderering of the authentication prompt
  - Improved support for monitor dis-/connection events
  - Improved lock screen display on lid open events
  - Improved idle activation
  - Improved low-power state, screensavers are now suspended
    when monitors are powered off via DPMS (Debian #921835)
- Translation Updates:
  Basque, Belarusian, Catalan, Czech, Danish, Galician, Lithuanian,
  Malay, Portuguese, Russian, Serbian, Spanish, Swedish

0.1.3 (beta)
============
- Features:
  - All available settings are now availble in the Preferences dialog
  - Locking on user inactivity is now based on X11 Screensaver signals
  - Xfce Screensaver now respects the xdg-screensaver state
  - Screensaver and lock screen functionality can now be toggled separately
- General:
  - Removed deprecated DBUS methods and signals from documentation
- Preferences:
  - Dropped unused preferences
  - Renamed all Xfconf properties for improved clarity
  - Embedded keyboard options have been disabled since they are currently
    non-functional
- Bug Fixes:
  - Replaced Help link with a link to the Xfce Documentation (Xfce #14877)
  - Added /usr/lib and /usr/libexec as trusted engine paths, enables
    local installs with access to existing screensavers (Xfce #14883)
  - Fixed screen blanking and locking on BSD (Xfce #14846)

src/xfce4-screensaver.c
/* -*- Mode: C; indent-tabs-mode: nil; c-basic-offset: 4; tab-width: 4 -*-
 *
 * Copyright (C) 2004-2006 William Jon McCann <mccann@jhu.edu>
 * Copyright (C) 2018      Sean Davis <bluesabre@xfce.org>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *
 * Authors: William Jon McCann <mccann@jhu.edu>
 *
 */

#include <config.h>

#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <gtk/gtk.h>

#include <libxfce4util/libxfce4util.h>
#include <xfconf/xfconf.h>

#include "gs-monitor.h"
#include "gs-debug.h"
#include "xfce4-screensaver.h"

void xfce4_screensaver_quit(void) {
    gtk_main_quit();
}

int main(int    argc,
         char **argv) {
    GSMonitor *monitor;
    GError    *error = NULL;

    static gboolean show_version = FALSE;
    static gboolean debug = FALSE;

    static GOptionEntry entries[] = {
        {"version", 0, 0, G_OPTION_ARG_NONE, &show_version, N_("Version of this application"), NULL},
        {"debug", 0, 0, G_OPTION_ARG_NONE, &debug, N_("Enable debugging code"), NULL},
        {NULL}
    };

    #ifdef ENABLE_NLS
        bindtextdomain(GETTEXT_PACKAGE, XFCELOCALEDIR);
        #ifdef HAVE_BIND_TEXTDOMAIN_CODESET
            bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
        #endif
        textdomain(GETTEXT_PACKAGE);
    #endif

    if (!gtk_init_with_args(&argc, &argv, NULL, entries, NULL, &error)) {
        if (error) {
            g_warning("%s", error->message);
            g_error_free(error);
        } else {
            g_warning("Unable to initialize GTK+");
        }

        exit(1);
    }

    if (show_version) {
        g_print("%s %s\n", argv[0], VERSION);
        exit(1);
    }

    if (!xfconf_init(&error)) {
        g_error("Failed to connect to xfconf daemon: %s.", error->message);
        g_error_free(error);

        return EXIT_FAILURE;
    }

    /* debug to a file if in deamon mode */
    gs_debug_init(debug, FALSE);
    gs_debug("Initializing xfce4-screensaver %s", VERSION);

    monitor = gs_monitor_new();

    if (monitor == NULL) {
        exit (1);
    }

    error = NULL;

    if (!gs_monitor_start(monitor, &error)) {
        if (error) {
            g_warning("%s", error->message);
            g_error_free(error);
        } else {
            g_warning("Unable to start screensaver");
        }

        exit(1);
    }

    gtk_main();

    g_object_unref(monitor);

    gs_debug("xfce4-screensaver finished");

    gs_debug_shutdown();

    return 0;
}
